### ISSUE TITLE
Building in manylinux container

### ISSUE THREAD

**Issue by andrewleech** [kayhayen/Nuitka#81]
Hi, 
I'm trying `bdist_nuitka` inside a manylinux container to create linux binary wheels for distribution:
https://github.com/pypa/manylinux

Have you ever tried this at all?

I get failures at the final linking step (with debug enabled), 
```
g++ -o mypkg.so -export-dynamic -Wl,-R,'/opt/python/cp35-cp35m/lib' -Wl,--disable-new-dtags 
-Wl,-b -Wl,binary -Wl,/build/build/lib/mypkg.build/__constants.bin -Wl,-b -Wl,elf64-x86-64 
-Wl,-defsym -Wl,constant_bin=_binary__build_build_lib_mypkg_build___constants_bin_start 
-shared mypkg.build/module.mypkg.common.finder.os mypkg.build/module.mypkg.helpers.os 
<<SNIP FILE LIST>>
mypkg.build/fibers_x64.os mypkg.build/swapfiber.os -L/opt/python/cp35-cp35m/lib 
-lpython3.5m -ldl -lpthread -lutil -lm

/opt/rh/devtoolset-2/root/usr/libexec/gcc/x86_64-CentOS-linux/4.8.2/ld: cannot find -lpython3.5m
```
Because the dynamic library is not included in the environment (by design?)
https://github.com/pypa/manylinux/issues/69
https://github.com/pypa/manylinux/issues/85

> I believe the correct fix is for the OpenCV build process to allow the Python symbols to be undefined at link time

http://vtk.1045678.n5.nabble.com/Fwd-OSX-build-links-directly-with-Python-binary-avoidable-td5731734.html

https://gitlab.kitware.com/vtk/vtk/merge_requests/1511

To be honest I'm a little lost with regard to how it could possibly build without linking against the library, but I gather it would not want to be locked into the library in the build environment, so that it can use the end-user's library instead?

I've also failed to find where scons gets its reference to `-lpython3.5m` from.  I've tried manually deleting it from both python3-config and the related pkg-config files and scons still finds it.

Considering the manylinux environment is specifically designed for building compatible binary wheels, shouldn't this be possible, even recommended?

**Comment by kayhayen** [comment]
There have been previous efforts to compile against a Python that was statically linked, I think for making binaries though. You want to make an extension module, right?

**Comment by andrewleech** [comment]
Yes it's for building pyd's for modules.
That's the thing, they don't seem to be referring to linking statically either: https://github.com/pypa/manylinux/issues/69#issuecomment-223768621

I'll have to ask them about it too, I feel like I must be missing something really obvious here.

**Comment by andrewleech** [comment]
Ok, yeah I finally found the place this is defined and turned it off when building a module (not on windows).
Really don't know if I've done it correctly, but it does build.

Unfortunately I don't have any way to actually test the module in question myself, that'll have to wait until next week now

**Comment by kayhayen** [comment]
The condition of the patch is horrible. It should be simple though, to detect if this Python is shared or not. Is there some docker image of manylinux, that I could try out?

Yours,
Kay

**Comment by andrewleech** [comment]
Yes I expected it was never going to be something to merge as is, it was just to record what I'd tested.

And yes, manylinux is designed to be used with docker:
https://github.com/pypa/manylinux
https://quay.io/repository/pypa/manylinux1_x86_64
docker: quay.io/pypa/manylinux1_x86_64

**Comment by kayhayen** [comment]
Just looked at manylinux, seems like it actually is relevant if Nuitka works with it. I was thinking fringe distribution, no relevance. But it seems to be actually intended to be the ideal platform for (binary) wheel production.

Please allow some time before I get to it, but I will.

**Comment by kayhayen** [comment]
My docker skills are a bit rusty, but I managed to login into that image, and first thing I try is this:

[root@733b154a65e3 /]# /opt/python/cp27-cp27mu/bin/python-config --libs
-lpython2.7 -lpthread -ldl -lutil -lm

So, I was thinking this wouldn't contain a reference to the non-existing library (I didn't find it yet). It's normal to not show up in ldd `which python`, so how would the wheel build itself know? I am a bit at loss here.

**Comment by kayhayen** [comment]
Andrew, please try out the factory branch or this: http://nuitka.net/doc/factory.html

**Comment by andrewleech** [comment]
Thanks for the investigations, that sounds really good. I'll hopefully get a chance to test it today. 

I understand it'll never work for compiled programs (rather then modules). Someone's raised that before with the manylinux people and the response was basically that's by design. The manylinux project is just about compiling modules and they want the reliability of that to be most important, so will avoid changes for any other non module support

**Comment by kayhayen** [comment]
@andrewleech Do you know, if the output of python-config --ldflags improved in the mean time?

I have now Issue #93 and #94 in which it seems that at least in one instance, a statically linked Python does have a shared library nearby. The detection code falls back to this command, but that would regress this issue I figure now. The manylinux would be detected as non-static due to having -lpython2.7 in there.

Do I really want to get in the business of searching if a dynamic library exists?

**Comment by andrewleech** [comment]
Perhaps you could check if the library actually exists pre-compilation. 

Can ldconfig give the required info? https://serverfault.com/a/513938/346207

Or, if it sounds more like manylinux is the corner case... perhaps could inefficiently detect the linker failure the original way, then retry the newer static way.

**Comment by kayhayen** [comment]
I think my question was posed, when I was still thinking I need a static python detection that works.  Seems not, so no danger of regressing this issue.
